internal/cephfs/fsjournal.go

```diff
@@ -179,3 +246,132 @@ func reserveVol(ctx context.Context, volOptions *volumeOptions, secret map[strin
 
 	return &vid, nil
 }
+
+// reserveSnap is a helper routine to request a UUID reservation for the CSI SnapName and,
+// to generate the snapshot identifier for the reserved UUID.
+func reserveSnap(ctx context.Context, volOptions *volumeOptions, parentSubVolName string, snap *cephfsSnapshot, cr *util.Credentials) (*snapshotIdentifier, error) {
+	var (
+		vid       snapshotIdentifier
+		imageUUID string
+		err       error
+	)
+
+	j, err := snapJournal.Connect(volOptions.Monitors, cr)
+	if err != nil {
+		return nil, err
+	}
+	defer j.Destroy()
+
+	imageUUID, vid.FsSnapshotName, err = j.ReserveName(
+		ctx, volOptions.MetadataPool, util.InvalidPoolID,
+		volOptions.MetadataPool, util.InvalidPoolID, snap.RequestName,
+		snap.NamePrefix, parentSubVolName, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// generate the snapshot ID to return to the CO system
+	vid.SnapshotID, err = util.GenerateVolID(ctx, volOptions.Monitors, cr, volOptions.FscID,
+		"", volOptions.ClusterID, imageUUID, volIDVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	util.DebugLog(ctx, "Generated Snapshot ID (%s) for request name (%s)",
+		vid.SnapshotID, snap.RequestName)
+
+	return &vid, nil
+}
+
+// undoSnapReservation is a helper routine to undo a name reservation for a CSI SnapshotName.
+func undoSnapReservation(ctx context.Context, volOptions *volumeOptions, vid snapshotIdentifier, snapName string, cr *util.Credentials) error {
```

Comment:
this one can also share new helpers with `undoVolReservation()`

Reply:
same as above, more less vol and snap is treated differently.